nsacyber/HIRS
# [#863] Placed Dependencies and Dependency Versions In TOML file

### Description
I've observed that many dependencies in our build.gradle and settings.gradle files have hard-coded version numbers. To streamline the process of managing these versions, this PR consolidates all the dependencies in one easy-to-access file.

### Test Instructions using the Gradle build command:
1. Run the `./gradlew clean --refresh-dependencies build` command and verify that not only does this command pull all of the new dependencies but it also builds the entire application without any issues.

### Test Instructions using the RPM/Debian file:
1.  Another way to confirm that these changes are good is by building an RPM and installing then running the HIRS-ACA service (building the RPM requires building and assembling all the artifacts and running the ACA service locally).

2. Stop the current HIRS-ACA runing on your device (assuming you already have one running)
```
sudo systemctl stop hirs-aca
```
<br>

3. Get rid of the old HIRS-ACA service:

On Rhel Linux distros:
```
sudo dnf remove  HIRS_AttestationCA.x86_64
```

or on a Debian-based Linux distribution:

```
sudo  apt remove HIRS_AttestationCA.x86_64
```
<br>

5. Go to your HIRs repo and checkout this branch,:

```
cd HIRS/

git pull

git checkout https://github.com/nsacyber/HIRS/tree/v3_issue_863-write-dependencies-version-numbers-in-gradleproperties-file
```
<br>

7. Create a RPM/Debian file:

On Rhel-based distros:
```
./gradlew clean buildRPM
```
<br>


On Debian-based distros:
```
./gradlew clean buildDeb
```
<br>

8. Install the service:

On Rhel-based distros:
```
sudo dnf install HIRS*rpm
```

On Debian-based distros:
```
sudo dnf install HIRS*deb
```
<br>

9. Go to the HIRS website and verify that it all works fine.
```
https://localhost:8443/
```
<br>

### Summary Of Updates:

- Upgraded Gradle from **version 8.3** to **version 8.11.1**.
- Moved repetitive tasks to the root build.gradle file.
- Upgraded spring-framework-boot plugin from **version 3.0.6** to **version 3.0.13**.
- Upgraded spring-dependency-management plugin from **version 1.1.0** to **version 1.1.7**.
- Upgraded ospackage plugin from **version 11.4.0** to **version 11.10.0**.
- Upgraded spotbugs plugin from **version 6.0.4** to **version 6.0.13**.
- Upgraded gradle-jaxb plugin from **version 5.1.0** to **version 7.0.1**.
- Deleted unused dependencies and/or replaced dependencies that were already included in a bigger dependency (this was to ensure that we do not have any dependency mismatches).
- Placed all dependencies and dependencies versions in a toml file. Click on the following link for more details on Version Cataloguing: [Medium Article on Version Catalog](https://callmeryan.medium.com/gradle-version-catalog-728111fa210f) and see the table below to see all the changes:

### Dependencies Changes:
_(Certain dependencies have been omitted if they have been replaced by one of 
the below dependencies or if their versions haven't changed)_

Name | Old Version | New Version | Notes |
----- | ----- | ----- |  ----- | 
org.apache.httpcomponents:httpclient | N/A | 4.5.14 | Replaced vulnerable apache transitive dependency in **com.github.marandus:pci-ids**. |
org.bouncycastle:bcmail-jdk18on | 1.77 | 1.79 | 
commons-codec:commons-codec | 1.15 | 1.17.1|
commons-io:commons-io | 2.11.0 | 2.18.0 |
org.apache.commons:commons-lang3 | 3.13.0 | 3.17.0 |
com.google.code.gson:gson | 2.10.1 | 2.11.0 |
org.glassfish:jakarta.json | N/A| 2.0.1 |  Replaced the older **org.glassfish:javax.json** dependency with this one. |
org.glassfish.jaxb:jaxb-runtime | 2.3.1  | 4.0.5 |
com.google.guava:guava | 31.1-jre | 33.3.1-jre |
com.fasterxml.jackson.core:jackson-core | 2.14.2 | 2.18.2 |
com.fasterxml.jackson.core:jackson-databind | 2.14.2 | 2.18.2|
jakarta.persistence:jakarta.persistence-api | 3.1.0 | 3.2.0|
jakarta.xml.bind:jakarta.xml.bind-api | 4.0.0 | 4.0.2 |
org.jcommander:jcommander | 1.83  | 2.0 |
org.projectlombok:lombok | 1.18.26 | 1.18.36|
org.mariadb.jdbc:mariadb-java-client | 3.1.4 |  3.5.1 |
com.google.protobuf:protobuf-java | 3.24.1 | 4.28.3 |
org.slf4j:slf4j-simple | 1.7.30  | 2.0.16 |
org.springframework.boot:spring-boot-starter-data-jpa | 3.0.1 | 3.4.0 |
org.springframework.boot:spring-boot-starter-log4j2 | 3.0.1 | 3.4.0 |
org.springframework.boot:spring-boot-starter-tomcat | 3.0.1 | 3.4.0 |
org.springframework.boot:spring-boot-starter-validation | 3.0.1 | 3.4.0 |
org.springframework.retry:spring-retry | 2.0.0 | 2.0.10 |
org.springframework.boot:spring-boot-starter-web  | 3.0.1 | 3.4.0 |
org.springframework:spring-webmvc | N/A | 6.2.1  | Replaced vulnerable webmvc dependency in **org.springframework.boot:spring-boot-starter-web**.
org.apache.tomcat.embed:tomcat-embed-jasper | 10.1.5 | 10.1.33 |
org.springframework.boot:spring-boot-starter-test  | 3.0.1 | 3.4.0 | Junit, Mockito, and Hamcrest, and testng have all been removed since these testing dependencies are packaged with the spring-boot-starter test dependency.

### Issues this PR addresses:
Closes #863
